shared/source/os_interface/linux/xe/ioctl_helper_xe.cpp

```diff
@@ -791,10 +819,128 @@ bool IoctlHelperXe::setVmSharedSystemMemAdvise(uint64_t handle, const size_t siz
     }
     vmIdsStr += "]";
     xeLog(" -> IoctlHelperXe::%s h=0x%x s=0x%lx vmids=%s\n", __FUNCTION__, handle, size, vmIdsStr.c_str());
-    // There is no vmAdvise attribute in Xe, so return success
+
+    drm_xe_madvise vmAdvise{};
+
+    vmAdvise.vm_id = 0;
+    vmAdvise.start = alignDown(handle, MemoryConstants::pageSize);
+    vmAdvise.range = alignSizeWholePage(reinterpret_cast<void *>(handle), size);
+    vmAdvise.type = attribute;
+
+    if (attribute == this->getPreferredLocationAdvise()) {
+        uint32_t devmen_fd = static_cast<uint32_t>(param >> 32);
+        uint32_t migrationPolicy = static_cast<uint32_t>(param & 0xFFFFFFFF);
+        vmAdvise.preferred_mem_loc.devmem_fd = devmen_fd;
+        vmAdvise.preferred_mem_loc.migration_policy = migrationPolicy;
+    } else if (attribute == this->getAtomicAdvise(false)) {
+        uint32_t val = static_cast<uint32_t>(param >> 32);
+        vmAdvise.atomic.val = val;
+    } else {
+        return false;
+    }
+
+    for (auto vmId : vmIds) {
+
+        // Call madvise on all VM Ids.
+        vmAdvise.vm_id = vmId;
+        auto ret = IoctlHelper::ioctl(DrmIoctl::gemVmAdvise, &vmAdvise);
+
+        xeLog(" vm=%d start=0x%lx size=0x%lx param=0x%lx operation=%d(%s) ret=%d\n",
+              vmAdvise.vm_id,
+              vmAdvise.start,
+              vmAdvise.range,
+              param,
+              vmAdvise.type,
+              xeGetAdviseOperationName(vmAdvise.type),
+              ret);
+
+        if (ret != 0) {
+            xeLog("error: %s ret=%d\n", xeGetAdviseOperationName(vmAdvise.type), ret);
+            return false;
+        }
+    }
+
     return true;
 }
 
+AtomicAccessMode IoctlHelperXe::getVmSharedSystemAtomicAttribute(uint64_t handle, const size_t size, const std::vector<uint32_t> &vmIds) {
+    std::string vmIdsStr = "[";
```

Comment:
do not create string just for purpose of logging
this would cause overhead even if logging is disabled